### PR TITLE
CDAP-4409 Script Context

### PR DESCRIPTION
**Pull request by awholegunch**
Fix for https://issues.cask.co/browse/CDAP-4409 and https://issues.cask.co/browse/CDAP-4194

Passes [Quick Build 3](http://builds.cask.co/browse/CDAP-DOB170-3)
Updated to match current changes in the Hydrator repo.

Builds the Validator function table from the Markdown file downloaded from the Hydrator repo.

Pages of interest:

- [Custom Plugins: Script Transformations](http://builds.cask.co/artifact/CDAP-DOB170/shared/build-3/Docs-HTML/3.4.0-SNAPSHOT/en/cdap-apps/hydrator/custom.html#script-transformations)
- [CoreValidator documentation](http://builds.cask.co/artifact/CDAP-DOB170/shared/build-3/Docs-HTML/3.4.0-SNAPSHOT/en/cdap-apps/hydrator/hydrator-plugins/shared-plugins/core.html)
- [Validator Transform](http://builds.cask.co/artifact/CDAP-DOB170/shared/build-3/Docs-HTML/3.4.0-SNAPSHOT/en/cdap-apps/hydrator/hydrator-plugins/transforms/validator.html)

### CODE REVIEW

File: cdap-docs/cdap-apps/source/hydrator/custom.rst [hunk 1]
```diff
@@ -363,6 +364,66 @@ copies in each transform is emitted. The user metrics can be queried by using th
...
+Currently, ``ScriptContext.getLookup(String table)`` only supports :ref:`key-value tables <datasets-index>`.
+
+For example, if a lookup table *purchases* is configured, then you will be able to perform
+operations with that lookup table in your script: ``context.getLookup('purchases').lookup('key')``
+
+**Validator Object**
+
+For example, in a validator transform, you can get the validator object and then call its functions as part of your script::
+
+  var coreValidator = context.getValidator(\"coreValidator\");
```
Comment (anwar6953) [inline cdap-docs/cdap-apps/source/hydrator/custom.rst]: the backslash is necessary?

Comment (awholegunch) [inline cdap-docs/cdap-apps/source/hydrator/custom.rst]: yes and no...this is copied and is out-of-context. It's actually JavaScript embedded in JSON, so it does actually need the backslashes when it is in the JSON. But, it's mis-leading in this case. I will remove them, and point out that this is JavaScript.

File: cdap-docs/tools/docs-extract-table.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                if in_table and not in_header and line.startswith(TABLE_RULE):
+                    in_table = False
+                    break                    
+    else:
+        print "Did not find %s" % input_file
+        print "Wrote 'missing file' to table"
+        lines.append(MISSING_FILE_TEMPLATE % input_file)
+        return_code = 2
+            
+    output = open(output_file,'w')
```
Comment (anwar6953) [inline cdap-docs/tools/docs-extract-table.py]: line 141 is redundant. Not sure if opening the same file for writing would cause any weird issues.

Comment (awholegunch) [inline cdap-docs/tools/docs-extract-table.py]: Fixed. Good catch.

File: cdap-docs/tools/docs-extract-table.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    comma = ','
+    return comma.join(pieces)
+
+def create_from_lines(line1, line2):
+    """ Combines two lines from a Markdown table and converts them into a single line for a reST CSV table line.
+    """
+    pieces1 = [piece.strip() for piece in line1.split(PIPE)][1:-1]
+    pieces2 = [piece.strip() for piece in line2.split(PIPE)][1:-1]
+    
+    if len(pieces1) != len(pieces2):
```
Comment (anwar6953) [inline cdap-docs/tools/docs-extract-table.py]: is this an error case?

File: cdap-docs/tools/docs-extract-table.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        return ''
+        
+    pieces = []
+    for i in range(len(pieces1)):    
+        pieces.append(pieces1[i] + ' ' + pieces2[i])
+    
+    pieces = [piece.strip() for piece in pieces]
```
Comment (anwar6953) [inline cdap-docs/tools/docs-extract-table.py]: these 5 lines seem common between this function and the function above this.

Comment (anwar6953) [inline cdap-docs/tools/docs-extract-table.py]: Also, line 80 seems redundant, because we already strip the pieces in lines 70 and 71.

Comment (awholegunch) [inline cdap-docs/tools/docs-extract-table.py]: Fixed. Factored out duplicate code.

File: cdap-docs/tools/docs-extract-table.py (hunk 3 above)
Comment (awholegunch) [inline cdap-docs/tools/docs-extract-table.py]: Added an exception.

**Comment by anwar6953** [comment]
LGTM

File: cdap-docs/tools/docs-extract-table.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    else:
+        print "Did not find %s" % input_file
+        print "Wrote 'missing file' to table"
+        lines.append(MISSING_FILE_TEMPLATE % input_file)
+        return_code = 2
+            
+    with open(output_file,'w') as output:
+        output.write(TABLE_HEADER)
+        for line in lines:
+            output.write(line+'\n')
```
Comment (gsps1) [inline cdap-docs/tools/docs-extract-table.py]: you want to close this output stream after writing.

Comment (anwar6953) [inline cdap-docs/tools/docs-extract-table.py]: the `with open(...`
syntax in python automatically closes the file after the indentation is over.

Comment (gsps1) [inline cdap-docs/tools/docs-extract-table.py]: ok

**Comment by sreevatsanraman** [comment]
lgtm

**Comment by gsps1** [comment]
would be good to squash before merging. this PR had 8 commits..

**Comment by anwar6953** [comment]
Had already merged release/3.4 into this branch, which makes it difficult to squash.
Though, I agree, in the future, we should squash commits.
It makes looking back in the history to find causes of a bug easier.

**Comment by gsps1** [comment]
if he hadn't had any conflicts when merging release/3.4 into his branch, then while rebasing we can move his commits to the top and squash. that shouldn't be too bad.

**Pull request merged**